src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -1206,24 +1234,51 @@ protected void setAttribute(final short tag, final Object value) {
         setAttribute(tag, value, false);
     }
 
-    protected void setAttribute(final short tag, final Object value, final boolean isUnsignedArray) {
-        if (value != null &&
-                !(value instanceof Byte || value instanceof Short || value instanceof Integer ||
-                        value instanceof String || value instanceof Character || value instanceof Float ||
-                        value instanceof byte[] || value instanceof short[] || value instanceof int[] ||
-                        value instanceof float[])) {
-            throw new SAMException("Attribute type " + value.getClass() + " not supported. Tag: " +
-                    SAMTagUtil.getSingleton().makeStringTag(tag));
+    /**
+     * Checks if a long value is within the range of unsigned integer.
+     * @param value a long value to check
+     * @return true if within range and false otherwise
+     */
+    public static final boolean isValidUnsignedInteger(long value) {
+        return value >= 0 && value <= MAX_UINT;
+    }
+
+    /**
+     * Checks if the value is allowed as an attribute value.
+     *
+     * @param value the value to be checked
+     * @return true if the value is valid and false otherwise
+     */
+    public static final boolean isAllowedAttributeValue(final Object value) {
+        if (value instanceof Byte || value instanceof Short || value instanceof Integer ||
+                value instanceof String || value instanceof Character || value instanceof Float ||
+                value instanceof byte[] || value instanceof short[] || value instanceof int[] ||
+                value instanceof float[]) {
+            return true;
+        }
+
+        // a special case of unsigned integers:
+        if (value instanceof Long) {
+            return isValidUnsignedInteger((Long) value)
+                    || ((Long) value >= Integer.MIN_VALUE && (Long) value <= Integer.MAX_VALUE);
         }
```

Comment:
this is being sent to the wrong email address, sorry.

On Wed, Nov 11, 2015 at 10:53 AM, droazen <notifications@github.com> wrote:

> In src/java/htsjdk/samtools/SAMRecord.java
> <https://github.com/samtools/htsjdk/pull/339#discussion_r44547566>:
>
> > +     *
> > +     * @param value the value to be checked
> > +     * @return true if the value is valid and false otherwise
> > +     */
> > +    public static final boolean isAllowedAttributeValue(final Object value) {
> > +        if (value instanceof Byte || value instanceof Short || value instanceof Integer ||
> > +                value instanceof String || value instanceof Character || value instanceof Float ||
> > +                value instanceof byte[] || value instanceof short[] || value instanceof int[] ||
> > +                value instanceof float[]) {
> > +            return true;
> > +        }
> > +
> > +        // a special case of unsigned integers:
> > +        if (value instanceof Long) {
> > +            return isValidUnsignedInteger((Long) value)
> > +                    || ((Long) value >= Integer.MIN_VALUE && (Long) value <= Integer.MAX_VALUE);
> >          }
>
> Long is very much a special case, since the bam spec limits us to values
> that can fit in a int32_t or uint32_t. Bytes/Short/Integers can fit, but
> not necessarily Longs.
>
> @ronlevine <https://github.com/ronlevine> Have to ask that you abort this
> new review -- there were a specific set of changes requested in the last
> review, and if they have been met we should merge this as-is. I am away
> today on account of the holiday, but will be able to take a look late
> tonight.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/339/files#r44547566>.
>



-- 
Daniel Rozen